osclib/cache.py

```diff
@@ -0,0 +1,275 @@
+
+    The cache takes a list of regular expression patterns and time to live (ttl)
+    for API paths. In addition to the ttl the project context is taken into
+    account when available in order to expire all caches related to a project
+    when the remote server indicates a change was made more recently than the
+    local cache reflects. This provides a fairly robust cache that can handle
+    multiple users changing the same projects.
+
+    Cannot safely cache, for lengthy periods, paths that can update without user
+    interaction or that do not trigger the project updated timestamp to change.
+    Such paths include anything related to build status. When a source package
+    is updated the linked packages do not trigger an update of their project. As
+    such sources cannot be reliably cached for too long.
+
+    Any paths without a project context will be cleared when updated using this
+    cache, but obviously not for other contributors.
+    """
+
+    CACHE_DIR = os.path.expanduser('~/.cache/osc-plugin-factory')
+    TTL_LONG = 12 * 60 * 60
+    TTL_SHORT = 5 * 60
+    TTL_DUPLICATE = 3
+    PATTERNS = {
+        # Group members cannot be guaranteed, but change rarely.
+        '/group/[^/?]+$': TTL_SHORT,
+        # Clear target project cache upon request acceptance.
+        '/request/(\d+)\?.*newstate=accepted': TTL_DUPLICATE,
+        "/search/package\?match=\[@project='([^']+)'\]$": TTL_LONG,
+        # Potentially expire the latest_updated since it will be the only way to
+        # tell after an adi staging is removed. For now just cache the calls
+        # that occur in rapid succession.
+        "/search/project/id\?match=starts-with\(@name,'([^']+)\:'\)$": TTL_DUPLICATE,
+        # List of all projects may change, but relevant ones rarely.
+        '/source$': TTL_LONG,
+        # Sources will be expired with project, could be done on package level.
+        '/source/([^/?]+)(?:\?.*)?$': TTL_LONG,
+        # Project will be marked changed when packages are added/removed.
+        '/source/([^/]+)/_meta$': TTL_LONG,
+        '/source/([^/]+)/(?:[^/]+)/(?:_meta|_link)$': TTL_LONG,
+        # Handles clearing local cache on package deletes. Lots of queries like
+        # updating project info, comment, and package additions.
+        '/source/([^/]+)/(?:[^/?]+)(?:\?[^/]+)?$': TTL_LONG,
+        # Presumably users are not interweaving in short windows.
+        '/statistics/latest_updated': TTL_SHORT,
+    }
+
+    last_updated = {}
+
+    @staticmethod
+    def init():
+        Cache.patterns = []
+        for pattern in Cache.PATTERNS:
+            Cache.patterns.append(re.compile(pattern))
+
+        # Replace http_request with wrapper function which needs a stored
+        # version of the original function to call.
+        if not hasattr(osc.core, '_http_request'):
+            osc.core._http_request = osc.core.http_request
+            osc.core.http_request = http_request
```

Comment:
I left #625 as it provides the direct change approach. That is limiting in that methods in `osc.core` cannot be cached and requires a one line change to whatever calls are to be cached. Does the trick.